File: day 16/day16.R
dat <- readr::read_lines(file.path("day 16", "input.txt"))
# dat <- "s1,x3/4,pe/b" # nolint
# programs <- letters[c(1:5)] # nolint

partA <- FALSE
danceStream <- unlist(strsplit(dat, split = ","))

for (j in 1:(2 - partA)) {
  programs <- letters[c(1:16)]
  on <- 1
  progVec <- c()
  maxIter <- 10 ^ 9
  if (partA) {
    maxIter <- 1
  } else {
    if (exists("onRemainder")) {
      maxIter <- (maxIter / length(danceStream)) %% onRemainder
    } else {
      maxIter <- (maxIter / length(danceStream))
    }
  }
  while (on <= maxIter) {
    progVec <- c(progVec, paste(programs, collapse = "")) # nolint
    for (i in 1:length(danceStream)) {
      danceSplit <- unlist(strsplit(danceStream[i], split = ""))
      move <- danceSplit[1]
      moveProgs <- unlist(strsplit(paste(danceSplit[c(2:length(danceSplit))],
                                         collapse = ""), split = "/"))
      if (move == "s") {
        programs <- programs[c( ( (length(programs) + 1) - as.numeric(moveProgs)):length(programs),
                              1:(length(programs) - as.numeric(moveProgs)))]
      }
      if (move == "x") {
        swap1 <- programs[as.numeric(moveProgs[1]) + 1]
        swap2 <- programs[as.numeric(moveProgs[2]) + 1]
        programs[as.numeric(moveProgs[1]) + 1] <- swap2
        programs[as.numeric(moveProgs[2]) + 1] <- swap1
      }
      if (move == "p") {
        idx1 <- which(programs == moveProgs[1])
        idx2 <- which(programs == moveProgs[2])
        switch1 <- programs[which(programs == moveProgs[1])]
        switch2 <- programs[which(programs == moveProgs[2])]
        programs[idx1] <- switch2
        programs[idx2] <- switch1
      }
    }
    # find out after how many repetitions the programs are in same state
    if (paste(programs, collapse = "") %in% progVec) {
      onRemainder <- on
      cat("\nonRemainder is:", onRemainder, "\n")
      break
    }
    on <- on + 1
  }
}
cat("Final order:", paste(programs, collapse = ""))

File: day 16/day16.py
file = open('C:\Users\Besitzer\Tresor\john.locker\Mega\R working directory\\adventofcode2017\day 16\input.txt', 'r')
danceStream = file.readline().split(",")
# dat = "s1,x3/4,pe/b"

programs = ['a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p']

partA = True
maxIter = 10 ** 9
maxIter = maxIter / len(danceStream)
if partA:
    maxIter = 1

on = 1
while on <= maxIter:
  for i in range(len(danceStream)):
    danceSplit = list(danceStream[i])
    move = danceSplit[0]
    if "/" in danceStream[i]:
        moveProgs = danceStream[i][1:].split("/")
    else:
        moveProgs = int(danceStream[i][1:])
    if move == "s":
      programs = programs[(len(programs) - int(moveProgs)):len(programs)] + programs[0:(len(programs) - moveProgs)]
    if move == "x":
      swap1 = programs[int(moveProgs[0])]
      swap2 = programs[int(moveProgs[1])]
      programs[int(moveProgs[0])] = swap2
      programs[int(moveProgs[1])] = swap1
    if move == "p":
      idx1 = programs.index(moveProgs[0])
      idx2 = programs.index(moveProgs[1])
      switch1 = programs[idx1]
      switch2 = programs[idx2]
      programs[idx1] = switch2
      programs[idx2] = switch1
  on += 1
  if on % 1000 == 0:
      print "Percent done: " + str(on / 1000) + "%\n"
print "Final order: " + ''.join(programs)
